### ISSUE TITLE
Add __str__ and __repr__ to ConfigParser for Better Debugging and Unit Testing

### ISSUE THREAD

**Issue by Agent-Hellboy** [python/cpython#127011]
# Feature or enhancement

### Proposal:

The ConfigParser class supports dictionary-like operations, including the __contains__ protocol for checking section existence using the in operator. However, it lacks __str__ and __repr__ methods, essential for providing human-readable and developer-friendly representations of the configuration data. Adding these methods would enhance its usability, particularly for debugging, logging, and unit testing, by offering clear insights into its internal state.

current behavior:

```python
str_resp: <configparser.ConfigParser object at 0x7f57e7cc5160>
repr_resp: <configparser.ConfigParser object at 0x7f57e7cc5160>
```
suggested behavior:
```python

str_resp: {'GERRIT': {'api_key': 'val'}}
repr_resp: <ConfigParser: {'GERRIT': {'api_key': 'val'}}>
```

I haven’t started a thread on Python Discourse as this is a straightforward feature request requiring a simple accept or reject decision. I'd appreciate the clarification on whether there was a specific design reason for not including __str__ and __repr__. If approved, I will raise a PR.

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Linked PRs
* gh-127014
* gh-132966

**Comment by ZeroIntensity** [comment]
I think it makes sense to add a `__str__` and `__repr__`, but I'm not so sure about the proposed design. I'm fine with the proposed `__str__` (it's fine for users to confuse a `ConfigParser` for a dictionary because it's dict-like), but the `__repr__` might be problematic. Generally, the rule of thumb is that a `__repr__` should return an expression that creates the same object, so it should really display the documented initargs here rather than the contents (but there are exceptions to this rule all the time).

Go ahead and send a PR, and we can decide whether or not to go with it from there :)

**Comment by Agent-Hellboy** [comment]
>  the rule of thumb is that a __repr__ should return an expression that creates the same object, so it should really display the documented initargs here rather than the contents

I have used default values

**Comment by terryjreedy** [comment]
The main issue to me is that a ConfigParser can have 100s of entries.  Of course, the same is true of dicts ans other containers.  And unittest special cases susch builtins to condense error messages.

**Comment by Agent-Hellboy** [comment]
> The main issue to me is that a ConfigParser can have 100s of entries. Of course, the same is true of dicts ans other containers. And unittest special cases susch builtins to condense error messages.

Why not leave it as is? Users who want to print the output are likely dealing with small configuration files anyway. Alternatively, if necessary, I could truncate the dictionary representation with ... for larger files.

**Comment by brianschubert** [comment]
I don't think `ConfigParser` should be rendered exactly like a dict. That seems like too high a risk for confusion, especially since `ConfigParser` isn't fully compatible with `dict`/`Mapping`:

```python
import configparser

config = configparser.ConfigParser()
config.read_string("""
[foo]
a = 1
""")

>>> config.get("foo")
Traceback (most recent call last):
  File "<python-input-21>", line 1, in <module>
    config.get("foo")
    ~~~~~~~~~~^^^^^^^
TypeError: RawConfigParser.get() missing 1 required positional argument: 'option'
```
If you what you want is a convenient shorthand for debugging, then `ConfigParser._sections` may suit your needs better than a `__str__`/`__repr__` implementation:
```
>>> config._sections
{'foo': {'a': '1'}}
```
This gives you essentially the same result as the proposed `__str__` implementation, just without running the interpolations.

**Comment by Agent-Hellboy** [comment]
Hi @brianschubert 

The example shared highlights a difference between ConfigParser and dict, but it doesn’t invalidate the idea of enhancing ConfigParser for better debugging. I’m not suggesting that ConfigParser should fully mimic a dict—if that were the goal, it would have been inherited from Mapping. The comparison to dict was to emphasize usability and debugging convenience, not to imply complete parity.

Relying on private APIs like _sections for debugging isn’t ideal. While people can and do use private attributes occasionally, providing a Pythonic approach like improving __str__ or __repr__ would offer a clean, public way to inspect configurations. This doesn’t conflict with the existing API but complements it, enhancing usability for developers without changing its core behavior.

**Comment by jaraco** [comment]
I feel like if the parameters are included in the `__repr__`, there should be some methodology for deciding which parameters should be included and which not. The proposed approach only shows two seemingly arbitrarily-selected parameters of the 12 the constructor accepts. My preference would be to show all of the parameters, but only those that the caller supplied when constructing the parser. Of course, that would be more involved, especially if you want to detect if the caller explicitly supplied a default value (I don't think you need to do that).

Unfortunately, the most important summary information about the state isn't part of the constructor. For example, there's no indication if the configparser has loaded a file or files (or which ones), or how many sections or keys it might have. That information might be more useful to the user. You may also want to consider providing other state, like `(null)` (for prior to binding a file stream) or `(13 sections)`.

**Comment by ZeroIntensity** [comment]
IMO, this kind of style you suggested for the repr would work fine for the str:

```
<ConfigParser: {'GERRIT': {'api_key': 'val'}}>
```

It doesn't break any conventions since it's `__repr__`, clearly shows the content, and doesn't get it confused with a dictionary.

**Comment by Agent-Hellboy** [comment]
> It doesn't break any conventions since it's `__repr__`, clearly shows the content, and doesn't get it confused with a dictionary.

Kind of, but from the Python docs and [this StackOverflow post](https://stackoverflow.com/questions/1984162/what-is-the-purpose-of-repr-method-in-python), it says __repr__ should be used to get the object's representation, and __str__ is for the end user. I guess people like us are going to use __str__ only.

**Comment by ZeroIntensity** [comment]
Yes, that follows that, doesn't it? Users will see the `__str__` when printing and whatnot, and then the repr otherwise.

**Comment by serhiy-storchaka** [comment]
I do not think this is a good idea. ConfigParser is similar to file or database objects. The repr can contain some metainformation (some of the constructor arguments, but not all, they are numerous), but not the whole content.

You can add a simple helper in your program to get the full string representation:
```py
def to_str(cp):
    out = io.StringIO()
    cp.write(out)
    return out.getvalue()
```

**Comment by serhiy-storchaka** [comment]
`__str__` is added if the object is enough string-like, if we often prints it as a string or include it in format strings. This is why `__str__` is defined for enums -- because we always want to see them by name. Most classes do not define `__str__`, they fall back to `__repr__` for string representation.

`__repr__` is needed to distinguish the object from other objects. By default it contains the type name and the ID. It is enough, but ID is not convenient, especially for short-living objects. Collections can be distinguished by their content. Other classes contain the most important attributes.

ConfigParser is a complex case. On one side, it is a parser. It has attributes which specify how to parse config file. On other side, it is a collection, containing sections and key-value pairs. Different people can be interested to see different aspects of ConfigParser -- as a parser or as a collection. They need different representations. There is also a problem that small ConfigParsers, whose repr would fit in one or few lines, are very uncommon. People may need a multiline representation.

So, there is no single representation that would satisfy _everyone_, or even _most people_. People would continue to use their own code, for example `vars(cf)`, `dict(cf)` or `{k: list(v) for k, v in cf.items()}`, or something more complex.

**Comment by Agent-Hellboy** [comment]
Hi @serhiy-storchaka , Thanks for the detailed response, it helped me reason few things.

> ConfigParser is a complex case. On one side, it is a parser. It has attributes which specify how to parse config file. On other side, it is a collection, containing sections and key-value pairs. Different people can be interested to see different aspects of ConfigParser -- as a parser or as a collection. They need different representations. There is also a problem that small ConfigParsers, whose repr would fit in one or few lines, are very uncommon. People may need a multiline representation.

Maybe ConfigParser is purely a parser after all, every parser takes input and produces a representation. It only overlaps with a collection because it parses the configuration and returns a familiar data structure. The real issue is that it doesn’t provide a public API for validating its output in unit tests. When I wrote unittests for my configuration 7–8 months ago, I had no choice but to rely on methods like sections() and has_section(). I’m not sure why I didn’t just convert it to a dictionary. I originally proposed adding a to_str() method; perhaps I should instead propose a to_dict() method, since other parsing libraries include to_dict() specifically to make validation easier.

Btw, I’m still not clear whether you want __repr__ to include fewer initialization arguments, maybe even just the section names? or you doesn't want it? The suggestion wasn’t clear to me.

**Comment by merwok** [comment]
I agree with Serhiy here. ConfigParser is meant to be loaded from a file (or created), manipulated or inspected, and written to a file.

### `__str__`

What should a `__str__` method do? IMO it should be a way to convert an instance (configparser as a collection of data, not a parser) to its string serialization.. but we already have the `write` method for this, with `cp.write(file)` clearer to read than a hypothetical `print(cp, file=file)`.

When a string is needed, using StringIO is the common idiom.
On the other hand, modules like `json` have methods to read and write from/to files and strings, so there could be an argument for a method like `getvalue` here.
And given that we have the methods `read`, `read_file`, `read_string` and `read_dict` for input, it could be argued to add symmetric methods `as_string` and `as_dict`*

\* names taken in likeness to existing `astuple`, `as_tuple` and `_asdict` in the standard library; we don’t have `to_<type>`.

### `__repr__`

Here I agree that a good repr helps debugging, in unit tests («is this the actual object or the expected result?») and live code inspection («which config file is this?»).

A repr with the loaded filenames (or file objects / placeholders for string or dict inputs) and _maybe_ non-default `__init__` parameters (or more generally instance attributes, if these can be changed after `__init__` was called), would be more useful than the default. But I’m not sure about the parameters: `vars(cp)` is the general tool here!

**Comment by serhiy-storchaka** [comment]
> we don’t have `to_<type>`.

We have helper functions `tostring()` and `tostringlist()` in `ElementTree` which wrap the `write()` method. But we need to serialize an XML to a bytes object or string more often than to get a string representation of the config file, because they are often stored not in a file.

Although I admit that there are precedences of using `__str__()` or `__repr__()` for this: `robotparser`, `netrc`. But I do not think they are good examples.

**Comment by merwok** [comment]
OK, so we have `tostring` but my point still stands: not `to_str` or `to_dict`, rather `as_list`, `asdict`, `as_tuple`, `astuple`. (Email messages have `to_string` rather than `__str__` to pass parameters controlling the output.)

I propose again:
- add a simple `__repr__` with the list of parsed filenames (this is easy to extract from the linked PR)
- don’t include attributes, there is `vars` for this

I am not sure:
- value of adding `__str__` that dumps the data, given that we don’t round-trip cleanly (see #42811 and #130927)
- [edited] we could have the section names in the string? or the filenames like repr?
- value of adding `as_dict` to replace current idiom `{key: dict(val) for key, val in cp.items()}`

**Comment by Agent-Hellboy** [comment]
Hi @merwok 
Thanks for the review 
> value of adding as_dict to replace current idiom {key: dict(val) for key, val in cp.items()}

There should be a public API to validate the configuration; someone even suggested an access-protected attribute `_section` for unittest, really??? or maybe he misunderstood as debugging only. The object should expose a public api to validate the configuration. 

Again, people can argue that the public API isn't worth it, it's not able to validate the configuration states from where it's loaded, and all that, and I have no counterargument. As done earlier, stating the identity of configparse, whether it's a parser or a collection. 


One more thing, as people here are suggesting using internal APIs countring https://github.com/python/cpython/issues/127011#issuecomment-2866249460 to test my `routine under test`, why should I rely on your implementation details to test my configuration? Isn't it a test antipattern and code smell? Maybe he is suggesting implementing a fake. I have no idea how it can be a fake.

**Comment by merwok** [comment]
I don’t really understand what a dict-conversion method (replacing an existing one-line idiom!) had to do with validation, can you explain?

[edit] I think you mean: get the configparser contents as a dict for easier unit tests. OK, that’s one use case.

I also don’t understand the reference to internal APIs. `items` and `write` are part of ConfigParser’s interface.

**Comment by Agent-Hellboy** [comment]
> I also don’t understand the reference to internal APIs. `items` and `write` are part of ConfigParser’s interface.

Ohh, I quoted the wrong comment thread. Others have suggested I should use implementation details for unittest. I mean, his suggestion of using the reference code block doesn't qualify as a valid test double used throughout the industry; it's not a fake, maybe a fixture, but too complex for a fixture, maybe, I can't argue though.

**Comment by merwok** [comment]
I agree about not recommending accessing internal data `_sections`.

I think the mentions of double, fake or fixture are not relevant to our discussion, but don’t want to get derailed to debate that.

Now I will be silent for a few days to see what other core devs think 🙂
